op-e2e: Add ability to deploy dispute game contracts and test that games can be resolved

ajsutton opened the pull request:
**Description**

Introduces e2e helpers for the dispute game factory and fault dispute game.  The factory helper deployed the required contracts when created and configures the fault dispute game implementation.

Theoretically the deployment be done in the L1 genesis creation along with other redeploys but it was quite complex to make work. In particular the `FaultDisputeGame` contract needs the address of the `AlphabetVM` contract to be passed in as a constructor but the VM isn't deployed with a proxy and so doesn't have a fixed address.  Making the `DisputeGameFactory.setImplementation` call to configure the fault dispute game also never worked for me - consistently returning "no code at address" even right at the end of genesis configuration. It's likely possible to make it work but most tests don't need these contracts and there's plans to replace the genesis generation with forge scripts that are already deploying the dispute game contracts correctly. One advantage is that the dispute game length is now configurable per-test.

The helpers provide a high level API for tests to work with the dispute game contracts while encapsulating the specific actions.

Also modifies setup so that a system can be deployed with only the L1 chain and no L2 nodes.

**Tests**

The time travel test has now been updated to deploy the dispute game contracts, create a game, time travel and check it can be resolved.

**Metadata**

- https://linear.app/optimism/issue/CLI-4133/e2e-tests

ajsutton committed:
op-e2e: Add ability to deploy dispute game contracts and test that games can be resolved.

ajsutton committed:
op-e2e: Add ability to deploy dispute game contracts and test that games can be resolved.

ajsutton commented on op-e2e/e2eutils/disputegame/helper.go:
```diff
@@ -0,0 +1,102 @@
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+	trace := fault.NewAlphabetProvider(claimedAlphabet, 4)
+	rootClaim, err := trace.Get(2 ^ alphabetGameDepth - 1)
```
pfft, what a dumb language.... :P

refcell commented on op-e2e/e2eutils/disputegame/deploy.go:
```diff
@@ -0,0 +1,60 @@
+
+// deployDisputeGameContracts deploys the DisputeGameFactory, AlphabetVM and FaultDisputeGame contracts
+// It configures the alphabet fault game as game type 0 (faultGameType)
+// If/when the dispute game factory becomes a predeployed contract this can be removed and just use the
```
nbd, but this looks like a hanging sentence

refcell reviewed (approved):
Very nice, LGTM

ajsutton committed:
op-e2e: Add helper to run op-challenger in e2e tests (#6366)

Enhance test by running a challenger that defends the output and confirm it wins.

trianglesphere commented on op-e2e/e2eutils/disputegame/helper.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+var alphaExtraData = common.Hex2Bytes("1000000000000000000000000000000000000000000000000000000000000000")
+var alphabetVMAbsolutePrestate = uint256.NewInt(140).Bytes32()
```
This was my error. The character "'" should be decimal 96. The value "140" is in octal.

trianglesphere replied:
https://github.com/ethereum-optimism/optimism/pull/6369

ajsutton replied:
Blind copy/paste. :D Fun fact though - it doesn't make any difference because it's consistent with both the contracts and the challenger. I have changed it for consistency with expectations though.

trianglesphere reviewed (commented):
Just the one pre-state issue which was my fault. Everything else looks good.

ajsutton committed:
op-e2e: Fix alphabet absolute prestate.

ajsutton committed:
Merge branch 'develop' into aj/deploy-dispute-game-e2e

ajsutton committed:
Merge remote-tracking branch 'origin/develop' into aj/deploy-dispute-game-e2e

ajsutton committed:
op-e2e: Expect the challenger to resolve the game

mergify[bot] committed:
Merge branch 'develop' into aj/deploy-dispute-game-e2e
